### DISCUSSION TITLE
Use of `raise NotImplementedError` instead of `NotImplemented`

### DISCUSSION THREAD

**Discussion by nydelzaf** [sqlalchemy/sqlalchemy#10645] [Usage Questions]
I am here to discuss about an issue that we came across while triaging your project.

## Background 
While traiging your project, our bug fixing tool generated the following message(s)-

In file: [associationproxy.py](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/ext/associationproxy.py#L1980), class: _AssociationSet, there is a special method `__ixor__` that raises a NotImplementedError. If a special method supporting a binary operation is not implemented it should return [NotImplemented](https://docs.python.org/3/library/constants.html#NotImplemented). On the other hand, [NotImplementedError](https://docs.python.org/3/library/exceptions.html#NotImplementedError) should be raised from abstract methods inside user defined base classes to indicate that derived classes should override those methods. iCR suggested that the special method `__ixor__` should return NotImplemented instead of raising an exception. An example of how NotImplemented helps the interpreter support a binary operation is [here](https://docs.python.org/3/library/numbers.html#implementing-the-arithmetic-operations).

The same message is generated for `__imul__`, `__iand__`, `__isub__`, `__ior__` methods too.

## Details
The reason behind suggesting this fix is that developers often confuse `NotImplemented` and `NotImplementedError` equivalent, which is wrong. Using `NotImplemented` allows the Python interpreter to take over and try performing that operation. If it fails to do so, the interpreter throws a `TypeError`.

## Suggested Changes
It is suggested that `raise NotImplementedError` be replaced with `return NotImplemented`. If you're willing, I can create a fix and submit a PR.

## Previously Found & Fixed
- https://www.github.com/SciTools/iris/pull/5544
- https://www.github.com/cupy/cupy/pull/7900
- https://www.github.com/ethereum/web3.py/pull/3080

## Sponsorship and Support
This work is done by the security researchers from OpenRefactory and is supported by the [Open Source Security Foundation (OpenSSF)](https://openssf.org/): [Project Alpha-Omega](https://alpha-omega.dev/). Alpha-Omega is a project partnering with open source software project maintainers to systematically find new, as-yet-undiscovered vulnerabilities in open source code - and get them fixed – to improve global software supply chain security.

The bug is found by running the Intelligent Code Repair (iCR) tool by OpenRefactory and then manually triaging the results.

**Comment by zzzeek** [comment]
hi -

Looking at the implementation of these methods, they are all actually implemented.  it appears the use of NotImplementedError is actually for the purposes of testing that the incoming argument is of the expected type.  I would gather that the history here involved anticipation of other set-like types that the methods didn't know how to handle, but that was decades ago.

Returning `NotImplemented` indicates that the operation should proceed in terms of other methods, which is not something we need to do here.  These methods are implemented, so that is definitely not the correct change.

These methods, to the degree that they are checking input and raising exceptions, would more correctly be converted to `TypeError`, as they are attempting to flag incorrect input types.  However, we are not doing this within the non-dundered versions of these methods, nor do we have test suite support for unsupported types being passed in, so I think the correct solution is these conditional raises of `NotImplementedError` should be removed completely and the method should be allowed to proceed given the input, in the same way as the non-dundered methods already work.

**Reply by nydelzaf** [reply]
Sorry, English isn't my first language. I didn't get the conclusion.

Are you suggesting -
1. `NotImplementedError` should be removed and replaced with something such as `TypeError`? or
2. Leave it at it is. No changes required.

Please help me understand. Thanks

**Reply by zzzeek** [reply]
the correct solution is these conditional raises of NotImplementedError should be removed completely and the method should be allowed to proceed given the input

**Reply by CaselIT** [reply]
I wonder why it was added. it has a check method and all `_set_binops_check_strict`

**Reply by nydelzaf** [reply]
Please check PR: https://github.com/sqlalchemy/sqlalchemy/pull/10648. If any changes are required, or if I need to create an issue, please let me know. Thanks.

**Reply by zzzeek** [reply]
revised here, based on the test run.    Using a list with the `__ior__` operator etc shoudl ultimately raise TypeError, so we should keep collections.py the way it is and adjust associationproxy accordingly and likely add tests

**Comment by zzzeek** [comment]
superseded by #11349
